qiskit_experiments/library/characterization/spectroscopy.py

```diff
@@ -0,0 +1,135 @@
+
+        return options
+
+    def __init__(
+        self,
+        qubit: int,
+        frequencies: Iterable[float],
+        backend: Optional[Backend] = None,
+        absolute: bool = True,
+        analysis: Optional[BaseAnalysis] = None,
+        **experiment_options,
+    ):
+        """A spectroscopy experiment where the frequency of a pulse is scanned.
+
+        Args:
+            qubit: The qubit on which to run spectroscopy.
+            frequencies: The frequencies to scan in the experiment, in Hz.
+            backend: Optional, the backend to run the experiment on.
+            absolute: Boolean to specify if the frequencies are absolute or relative to the
+                qubit frequency in the backend.
+            analysis: An instance of the analysis class to use.
+            experiment_options: Key word arguments used to set the experiment options.
+
+        Raises:
+            QiskitError: if there are less than three frequency shifts.
+
+        """
+        analysis = analysis or ResonanceAnalysis()
+
+        super().__init__([qubit], analysis=analysis, backend=backend)
+
+        if len(frequencies) < 3:
+            raise QiskitError("Spectroscopy requires at least three frequencies.")
+
+        self._frequencies = frequencies
+        self._absolute = absolute
+
+        self.set_experiment_options(**experiment_options)
+
+    def _set_backend(self, backend: Backend):
+        """Set the backend for the experiment and extract config information."""
+        super()._set_backend(backend)
+
+        self._dt = getattr(self.backend.configuration(), "dt", None)
+        constraints = getattr(self.backend.configuration(), "timing_constraints", {})
+        self._granularity = constraints.get("granularity", None)
+
+        if self._dt is None or self._granularity is None:
+            raise QiskitError(f"{self.__class__.__name__} needs both dt and sample granularity.")
+
+    @property
+    @abstractmethod
+    def _backend_center_frequency(self) -> float:
+        """The default frequency for the channel of the spectroscopy pulse.
+
+        This frequency is used to calculate the appropriate frequency shifts to apply to the
+        spectroscopy pulse as its frequency is scanned in the experiment. Spectroscopy experiments
+        should implement schedules using frequency shifts. Therefore, if an absolute frequency
+        range is given the frequency shifts need to be corrected by the backend default frequency
+        which depends on the nature of the spectroscopy experiment.
+        """
+
+    def _add_metadata(self, circuit: QuantumCircuit, freq: float, sched: pulse.ScheduleBlock):
+        """Helper method to add the metadata to avoid code duplication with subclasses."""
+
+        if not self._absolute:
+            freq += self._backend_center_frequency
+
+        circuit.metadata = {
+            "experiment_type": self._type,
+            "qubits": self.physical_qubits,
```

Comment:
What is the circuit metadata used for? Does it matter that for the resonator spectroscopy the qubit is not really part of the experiment?

Reply:
The `qubits` entry does not matter. Tests run fine without it. The only crucial part of the metadata is the `xval` since this is what is used in the curve analysis. I think we can tacitly interpret this as being the qubit for which we are doing readout spec. Alternatively we could override the `_add_metadata` function in `ResonatorSpectroscopy` to change this field but this sounds like a bit too much.